lightning/src/events/mod.rs

```diff
@@ -270,6 +270,41 @@ impl_writeable_tlv_based_enum!(InterceptNextHop,
 	};
 );
 
+/// The reason the payment failed. Used in [`Event::PaymentFailed`].
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PaymentFailureReason {
+	/// The intended recipient rejected our payment or the payment failed along the path.
+	/// A [`Event::PaymentPathFailed`] should have come before this.
+	RecipientRejected,
+	/// The user chose to abandon this payment by calling [`ChannelManager::abandon_payment`].
+	///
+	/// [`ChannelManager::abandon_payment`]: crate::ln::channelmanager::ChannelManager::abandon_payment
+	UserAbandoned,
+	/// We exhausted all of our retry attempts while trying to send the payment, or we
+	/// exhausted the `Retry::Timeout` if the user set one.
+	ExhaustedRetryAttempts,
+	/// The payment expired while retrying. If the user chooses to include an `expiry_time` in their
+	/// `PaymentParams` this payment will be failed if it is retried after this time has elapsed.
+	PaymentExpired,
+	/// We failed to find a route while retrying the payment.
+	FailedRoutingRetry,
+	/// The amount we would pay if we were to retry the payment would exceed the intended amount
+	/// above our allowed threshold.
+	RetryAmountOverflow,
+	/// Parameter error while sending with retry.
+	RetryParameterError,
```

Comment:
I’m thinking should I consolidate these different retry errors (or group them under their own enum?)? They communicate more information, but I’m not really sure how practically helpful they are to a user.